Fix some easy migration spec Rails 4+ warnings

jrafanie opened the pull request:
#### Build relations instead of using finder options.

    fixes warning: Relation#calculate with finder options is deprecated. Please build a scope and then call calculate on it instead.

#### Use Relation#to_a instead of #all when we want an array

    Fixes warning: Relation#all is deprecated. If you want to eager-load a relation, you can call #load (e.g. `Post.where(published: true).load`). If you want to get an array of records fro

####  Use existing timestamp behavior of allowing nulls (t.timestamp, t.timestamps)

    Fixes warning:
    `#timestamp` was called without specifying an option for `null`.
    In Rails 5, this behavior will change to `null: false`.
    You should manually specify `null: true` to prevent the behavior of your existing migrations from changing.

jrafanie commented:
cc @matthewd @tenderlove (note I couldn't get all of the timestamp warnings... see test output... :confused:)

I need your help/guidance how to fix the next noisy warning in these specs.

We use sanitize_sql_for_conditions with options and table name in the migrations... I believe we have to use the migration stub model to build relations, not sure if you have thoughts... :pray:   See [here](
https://github.com/jrafanie/manageiq/blob/75b4a9bac77699a55656aed1abd22877d11282ef/vmdb/lib/migration_helper.rb#L112), we call this all over this file.

If you want to work on these :wink:, you can see the warnings by running `bundle exec rake test:migrations:setup && bundle exec rake test:migrations` from vmdb.

jrafanie commented:
I'll fix the rubocop warning...

jrafanie commented:
All fixed now

kbrock commented on vmdb/db/migrate/20130809172032_change_miq_event_description_from_mgmt_sys_to_provider.rb:
```diff
@@ -14,7 +14,7 @@ def up
   def down
     say_with_time("Changing MiqEvent description with Provider to Mgmt Sys ") do
-      MiqEvent.where("description like ? and name like ?", "Provider%", "ems_auth_%").all.each do |e|
+      MiqEvent.where("description like ? and name like ?", "Provider%", "ems_auth_%").each do |e|
```
should these be `.to_a.each`, or just plain `.each` are good enough?

Fryguy replied:
.each is good enough given you have an AR::Relation.

Fryguy commented on vmdb/db/migrate/20100308005125_create_miq_cim_instances.rb:
```diff
@@ -14,7 +14,7 @@ def self.up
       t.column :stat_id,          :integer
       t.column :stat_top_id,        :integer
-      t.timestamps
+      t.timestamps :null => true
```
We are going to have to make bot rule for this...no one will remember this.

jrafanie replied:
Well, hopefully people are running them locally and will get the warning but yes, when rails5 comes and defaults to false and if the warning goes away, yes, you'll have no idea.

matthewd replied:
Why do we care?

jrafanie replied:
Yeah, good point... I don't know that any new code will rely upon the old way.  As long as we clean up all of the warnings on existing code now, we should be ok when the default changes.

Fryguy replied:
If we want everything to go the new way, then this PR should get merged, and there should be a data migration to fix the database to be accurate in the "new" way.

Fryguy commented on vmdb/db/migrate/20150111161446_create_nodes.rb:
```diff
@@ -3,7 +3,7 @@ def up
       t.string    :ems_ref
       t.string    :name
-      t.timestamp  :creation_timestamp
+      t.timestamp  :creation_timestamp, :null => true
```
Is the default for any timestamp column to be not-null?  or is it only for the timestamps method.  If so, why was this changed?  Seems so arbitrary. @tenderlove ?

jrafanie replied:
Note, even with my changes in this PR, I still get a few warnings of the same kind that I haven't tracked down yet.  Perhaps datetime or other methods raise the warning but mention timestamp.

tenderlove replied:
> Is the default for any timestamp column to be not-null? or is it only for the timestamps method.

It *should* be any timestamp method.

I think it's a stop gap towards Rails 5 upgrading.  In Rails 5, we're going to make timestamps default to not-null (make them default to NOW), but we can't do without forcing everyone to specify the default.  This goes back to conversations we were having about versioning migrations. /cc @matthewd

matthewd replied:
It's only for `timestamps` (the one that adds `created_at`/`updated_at`). It was made because it's impossible for those columns to actually contain NULL (because we always set them), so it's cleaner (and saves storage) to define the constraint.

A previous attempt (https://github.com/rails/rails/pull/3334) had to be reverted because it was an immediately breaking change. People still wanted it (e.g., https://github.com/rails/rails/pull/13885), so we found a way to do it through a deprecation cycle.

(Yes, versioned migration API would've made a much neater solution; this change was one of the main prompts that started us considering that for the future.)

Fryguy replied:
> It's only for timestamps

Ah, that makes sense then.  All timestamps would be really annoying.  Even so, we need to coordinate our migrations around upgrades as per my other comment.

matthewd replied:
https://github.com/rails/rails/commit/50d73692de4c8c492b84ad98cc07063bb74bab19 :+1:

Fryguy replied:
And this one.

jrafanie replied:
I don't understand, `timestamp` raises the warning too.  Are we saying the warning is wrong and only `timestamps` is going to change from `null: true` to `null: false`

matthewd replied:
> `timestamp` raises the warning too

I couldn't find a code path that would do that.

> Are we saying the warning is wrong

Yes, that's why I fixed it.

jrafanie replied:
Thanks, not sure why I missed your commit to fix it.  :+1: 

Ok, I'll rollback the changes to timestamp.  Thanks @matthewd

Fryguy commented on vmdb/spec/migrations/20120620215904_convert_old_states_to_new_format_spec.rb:
```diff
@@ -11,7 +11,7 @@
         disable_paralleism { migrate }
 
-        actual   = state_stub.order(:id).all.collect(&:data)
+        actual   = state_stub.order(:id).to_a.collect(&:data)
```
to_a Is not necessary here.

jrafanie replied:
Thanks for the sharp  👀

kbrock commented:
@jrafanie This looks great - thanks

jrafanie commented:
Fixed the uneeded to_a @Fryguy found.

Added issue to the bot to help us remember the change in the migration code for timestamps from null:true to null:false [here](https://github.com/ManageIQ/miq_bot/issues/69)

Anything else?

Fryguy commented on vmdb/db/migrate/20100428172512_create_miq_alert_statuses.rb:
```diff
@@ -4,7 +4,7 @@ def self.up
       t.integer     :resource_id
       t.string      :resource_type
-      t.timestamp   :evaluated_on
+      t.timestamp   :evaluated_on, :null => true
```
@jrafanie Can you undo this change?

Fryguy commented on vmdb/db/migrate/20141020195642_create_orchestration_stacks.rb:
```diff
@@ -60,7 +60,7 @@ def change
       t.string :resource_status
       t.text   :resource_status_reason
-      t.timestamp :last_updated
+      t.timestamp :last_updated, :null => true
```
and this one?

Fryguy commented on vmdb/db/migrate/20150111161504_create_kubernetes_services.rb:
```diff
@@ -3,7 +3,7 @@ def up
       t.string     :ems_ref
       t.string     :name
-      t.timestamp  :creation_timestamp
+      t.timestamp  :creation_timestamp, :null => true
```
Same

Fryguy commented on vmdb/db/migrate/20150128161824_create_container_groups.rb:
```diff
@@ -3,7 +3,7 @@ def up
       t.string     :ems_ref
       t.string     :name
-      t.timestamp  :creation_timestamp
+      t.timestamp  :creation_timestamp, :null => true
```
One more

Fryguy commented on vmdb/db/migrate/20150319115040_create_replication_controllers.rb:
```diff
@@ -3,7 +3,7 @@ def up
       t.string     :ems_ref
       t.string     :name
-      t.timestamp  :creation_timestamp
+      t.timestamp  :creation_timestamp, :null => true
```
:dizzy:

Fryguy commented:
I was gonna list them all, but there were way more than I expected :smile:

jrafanie commented:
Ok, @matthewd @Fryguy all done... please review...

tenderlove commented:
:+1: I will merge it.
